r2dbc-spi/src/main/java/io/r2dbc/spi/Result.java

```diff
@@ -49,4 +54,124 @@
+     * is subject to the {@code mappingFunction}.
+     *
+     * @param mappingFunction the {@link Function} that maps a {@link Segment} a to a value
+     * @param <T>             the type of the mapped value
+     * @return a mapping of the segments that are the results of a query against a database
+     * @throws IllegalArgumentException if {@code mappingFunction} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    <T> Publisher<T> flatMap(Function<Segment, ? extends Publisher<T>> mappingFunction);
+
+    /**
+     * Marker interface for a result segment.  Result segments represent the individual parts of a result from a query against a database.  Segments are typically update counts, data rows/{@code
+     * OUT} parameters, and messages of different severities.  Segments can also represent vendor-specific extensions.
+     *
+     * @since XXX
+     */
+    interface Segment {
+
+    }
+
+    /**
+     * Data segment consisting of {@link Row gettable data} and its {@link #metadata() RowMetadata}.  This segment can be backed by either a {@link Row} or any other type of gettable data such as
+     * {@code OUT} parameters.
+     *
+     * @since XXX
+     */
+    interface Data extends Segment, Row {
+
+        RowMetadata metadata();
+
+    }
+
+    /**
+     * Update count segment consisting providing a {@link #value() affected rows count}.
+     *
+     * @since XXX
+     */
+    interface UpdateCount extends Segment {
+
+        long value();
+
+    }
+
+    /**
+     * Message segment reported as result of the statement response.  Messages can vary in their {@link #severity() severity}.
+     *
+     * @since XXX
+     */
+    interface Message extends Segment {
+
+        /**
+         * Return the error as {@link R2dbcException}.
+         *
+         * @return the error as {@link R2dbcException}.
+         */
+        R2dbcException exception();
+
+        /**
+         * Returns the vendor-specific error code.
+         *
+         * @return the vendor-specific error code
+         */
+        int errorCode();
+
+        /**
+         * Returns the SQLState.
+         *
+         * @return the SQLState
+         */
+        @Nullable
+        String sqlState();
+
+        /**
+         * Returns the message text.
+         *
+         * @return the message text.
+         */
+        String message();
+
+        /**
+         * Returns the message severity.
+         *
+         * @return the message severity.
+         */
+        Severity severity();
+
+        enum Severity {
+
+            /**
+             * Any informal message.
+             */
+            INFO,
+
+            /**
+             * Error messages that terminate the statement result.
+             */
+            ERROR
```

Comment:
What about `WARN`? The current number of severities might be enough, but I think a discussion is worth having...

- In JDBC, we have `SQLWarning`, which is definitely more than "informal". 
- In PostgreSQL, there are various more levels: https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html

Notice that the claim of an `ERROR` terminating the statement result is probably not true for all RDBMS. There can be several chained errors, and it may be desirable to retrieve them all. In JDBC, this is possible through two ways:

1. Via `SQLException.getNextException()`
2. By calling `Statement.getMoreResults()` again after catching a `SQLException`, see: https://blog.jooq.org/2017/07/13/how-i-incorrectly-fetched-jdbc-resultsets-again/

Reply:
Severities are by no means exhaustive, it's more for the illustrative purpose, for now, to distinguish between server notifications that shall terminate result consumption (`via `map(…)` and `getRowsUpdated()`) and everything else.

The idea of `flatMap(…)` is, that one gets hold of the message, and then they can do whatever they want to do with it. Throwing an exception and terminating result consumption is just one of the options. Collecting all messages without terminating result consumption would be another possibility. Or ignoring these.

Reply:
> The idea of `flatMap(…)` is, that one gets hold of the message, and then they can do whatever they want to do with it. Throwing an exception and terminating result consumption is just one of the options. Collecting all messages without terminating result consumption would be another possibility. Or ignoring these.

I see. So, then this is merely a matter of changing the Javadoc for it not to establish such a contract.

Reply:
Revisiting this one before the merge. Since trying to press severities into one enumeration, how about rather expressing whether a message/error is formal or informal? Anything informal can map to a notification (`SQLWarning`) and formal messages are ones that terminate the statement processing (nature of exceptions).

Reply:
If there are doubts, I'm in favour of dropping it and revisiting the concepts later on. We already have the important bits of information, which are the error message and the sql state and error code.

If the enum doesn't capture the real world sufficiently, then it's quite hard to change later on.